Improper pairing of type and attribute for facet query causes 500

Low priority

```
curl -X POST "http://localhost:8080/api/metagenome_assembly/facet" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"conditions\":[{\"field\":\"id\",\"op\":\"==\",\"value\":\"nmdc:f452368cf4b2165458792d88709fddcc\",\"table\":\"metagenome_annotation\"}],\"attribute\":\"omics_type\"}"
```

FWIW, this query was an accident, that pairing of type (metagenome_assemby) and attribute (omics_type) should never have happened.  However, I think this should be a client error rather than a 500.

```
backend_1  | 172.18.0.1:43310 - "POST /api/metagenome_annotation/facet HTTP/1.1" 500
backend_1  | [2020-07-20 21:04:32 +0000] [23] [ERROR] Exception in ASGI application
backend_1  | Traceback (most recent call last):
backend_1  |   File "/usr/local/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 385, in run_asgi
backend_1  |     result = await app(self.scope, self.receive, self.send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
backend_1  |     return await self.app(scope, receive, send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/fastapi/applications.py", line 181, in __call__
backend_1  |     await super().__call__(scope, receive, send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/applications.py", line 102, in __call__
backend_1  |     await self.middleware_stack(scope, receive, send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
backend_1  |     raise exc from None
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
backend_1  |     await self.app(scope, receive, _send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
backend_1  |     raise exc from None
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
backend_1  |     await self.app(scope, receive, sender)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 550, in __call__
backend_1  |     await route.handle(scope, receive, send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 227, in handle
backend_1  |     await self.app(scope, receive, send)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 41, in app
backend_1  |     response = await func(request)
backend_1  |   File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 196, in app
backend_1  |     raw_response = await run_endpoint_function(
backend_1  |   File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 147, in run_endpoint_function
backend_1  |     return await dependant.call(**values)
backend_1  |   File "/app/nmdc_server/api.py", line 332, in facet_metagenome_annotation
backend_1  |     return crud.facet_metagenome_annotation(db, query.attribute, query.conditions)
backend_1  |   File "/app/nmdc_server/crud.py", line 220, in facet_metagenome_annotation
backend_1  |     facets = query.MetagenomeAnnotationQuerySchema(conditions=conditions).facet(db, attribute)
backend_1  |   File "/app/nmdc_server/query.py", line 327, in facet
backend_1  |     column = model.annotations[attribute]  # type: ignore
backend_1  | AttributeError: type object 'MetagenomeAnnotation' has no attribute 'annotations'
```

--- a/nmdc_server/errors.py
+++ b/nmdc_server/errors.py
@@ -4,10 +4,18 @@ from fastapi import FastAPI
 from starlette.requests import Request
 from starlette.responses import JSONResponse
 
+from .query import InvalidAttributeException
 from .utils import log_extras, logger
 
 
 def attach_error_handlers(app: FastAPI) -> None:
+    @app.exception_handler(InvalidAttributeException)
+    async def invalid_attribute_handler(req: Request, exc: Exception) -> JSONResponse:
+        exception_id = str(uuid.uuid4())
+        logger.exception("Unhandled exception, ID=%s.", exception_id, extra=log_extras(req))
+
+        return JSONResponse(status_code=400, content={"detail": {"msg": str(exc),}},)
+
     @app.exception_handler(Exception)
     async def internal_exception_handler(req: Request, exc: Exception) -> JSONResponse:
         exception_id = str(uuid.uuid4())

--- a/nmdc_server/query.py
+++ b/nmdc_server/query.py
@@ -20,6 +20,15 @@ EnvMediumAncestor = aliased(models.EnvoAncestor)
 EnvMediumTerm = aliased(models.EnvoTerm)
 
 
+class InvalidAttributeException(Exception):
+    def __init__(self, table: str, attribute: str):
+        self.table = table
+        self.attribute = attribute
+        super(InvalidAttributeException, self).__init__(
+            f"Attribute {self.attribute} not found in table {self.table}"
+        )
+
+
 def _join_envo(query: Query) -> Query:
     return (
         query.join(
@@ -247,7 +256,7 @@ class Condition(ConditionSchema):
             elif self.op == Operation.between:
                 value = cast(Tuple[schemas.AnnotationValue, schemas.AnnotationValue], self.value)
                 return and_(column >= value[0], column <= value[1])
-        if self.op == Operation.between:
+        if self.op == Operation.between and hasattr(model, "annotations"):
             value = cast(Tuple[schemas.AnnotationValue, schemas.AnnotationValue], self.value)
             return and_(
                 func.nmdc_compare(
@@ -259,10 +268,8 @@ class Condition(ConditionSchema):
             )
         if hasattr(model, "annotations"):
             json_field = model.annotations  # type: ignore
-        elif hasattr(model, "stats"):
-            json_field = model.extra  # type: ignore
         else:
-            raise Exception("Invalid field name")
+            raise InvalidAttributeException(self.table.value, self.field)
         return func.nmdc_compare(
             json_field[self.field].astext, self.op.value, self.value  # type: ignore
         )
@@ -323,8 +330,10 @@ class BaseQuerySchema(BaseModel):
         else:
             if attribute in model.__table__.columns:
                 column = getattr(model, attribute)
-            else:
+            elif hasattr(model, "annotations"):
                 column = model.annotations[attribute]  # type: ignore
+            else:
+                raise InvalidAttributeException(self.table.value, attribute)
 
         subquery = self.query(db).subquery()
         query = db.query(column, func.count(column))

--- a/tests/test_query.py
+++ b/tests/test_query.py
@@ -391,3 +391,17 @@ def test_pipeline_query(db: Session, table):
         conditions=[{"table": q.table.value, "field": "name", "value": f"{table}1"}]
     )
     assert ["project1"] == [r.name for r in q.execute(db).all()]
+
+
+def test_query_invalid_attribute(db: Session):
+    q = query.ReadsQCQuerySchema(
+        conditions=[{"table": "reads_qc", "field": "bad value", "value": ""}]
+    )
+    with pytest.raises(query.InvalidAttributeException):
+        q.execute(db)
+
+
+def test_facet_invalid_attribute(db: Session):
+    q = query.ReadsQCQuerySchema()
+    with pytest.raises(query.InvalidAttributeException):
+        q.facet(db, "bad value")

